### PR TITLE
[connectivity] No Location Permission preprocessor flag

### PR DESCRIPTION
**Pull request by KevinTheGray**
## Description

On iOS, if you do not specify `NSLocationAlwaysUsageDescription` or `NSLocationWhenInUseUsageDescription`, you can be rejected by iOS simply because `Connectivity` uses APIs that required them.  However, you may not wish to ever use this sensitive data, and only want to use `Connectivity` to check and listen for the user's reachability status.  This PR allows a user of the plugin to specify whether they would like to disable any code that uses these sensitive APIs by specifying a macro in their `Podfile`, similar to how the `permission_handler` plugin works.

I updated the README to explain how user's can utilize this.

## Related Issues

Resolves: https://github.com/flutter/flutter/issues/40418

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by kauaicreative** [comment]
I have read the updated the README. Are there instructions on, "specifying a macro in their Podfile"?

**Comment by KevinTheGray** [comment]
> are there instructions on, "specifying a macro in their Podfile"?

@kauaicreative, the README is modified to show how you specify a macro in the Podfile, particularly this bit:

```
To do so, add the following to your `Podfile`:

post_install do |installer|
  installer.pods_project.targets.each do |target|
    target.build_configurations.each do |config|
      ... # Here are some configurations automatically generated by flutter
      
      config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] ||= [
        '$(inherited)',
        'DISABLE_CONNECTIVITY_LOCATION_CODE=1'
      ]
    end
  end
end

To re-enable that code, you should remove the `'DISABLE_CONNECTIVITY_LOCATION_CODE=1'` and insure `pod install` is run again.
```

> I'm wondering if it safe to revert back to connectivity from simple_connectivity.

I am using this fork in my app with success and it is currently up to date with the plugins repo, but I think it may be better to stick with `simple_connectivity` until i get some feedback on this PR

**Comment by simoncent** [comment]
Would be great to get this merged. The `permission_handler` plugin is already using macros like this.

**Comment by stephenhuey** [comment]
Will it possibly take a while to get this approved?  And if so, @KevinTheGray do you feel pretty good about me going into production with your `simple_connectivity` version?

**Comment by KevinTheGray** [comment]
@stephenhuey I think it might be a little while, since it hasn't started review yet.  Once the review starts, I'm sure there will be some additional feedback that might take a bit to get done.  

I think `simple_connectivity` is safe for production, it looks like it offers everything `connectivity` does, without the location requirements.

**Comment by luismarquesfreire** [comment]
This issue is very critical since it doesn't allow us to submit app to App Store on iOS 14. I'm getting an error after App Store processing that I traced back to this - "ITMS-90842: Invalid SDK usage - App clip 'Runner.app/AppClips/AppClip.app' uses the SDK selector 'requestAlwaysAuthorization', which is not supported. Ensure your app clip is using supported SDK features."

File: packages/connectivity/connectivity/README.md
Comment (brunobowden) [inline packages/connectivity/connectivity/README.md:114]: Thanks for the proposal. Documentation should be updated due to https://github.com/flutter/flutter/pull/59044, see line 23 here:
https://github.com/flutter/flutter/pull/59044/files#diff-9a4a61b0e60143d9364139b944b2d485R23

```suggestion
post_install do |installer|
  installer.pods_project.targets.each do |target|
    flutter_additional_ios_build_settings(target)
    # Add this section:
    target.build_configurations.each do |config|
       config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] ||= [
        '$(inherited)',
        'DISABLE_CONNECTIVITY_LOCATION_CODE=1'
    ]
  end
end
```

File: packages/connectivity/connectivity/pubspec.yaml
Comment (brunobowden) [inline packages/connectivity/connectivity/pubspec.yaml:8]: As above, I think this should be at least 0.5 given the significance of the change.

**Review by brunobowden — COMMENTED** [review]
@KevinTheGray - I'd defer to @cyanglaz for his opinion on this but this is meant to be some helpful interim comments.

**Comment by brunobowden** [comment]
@KevinTheGray - please update here when you've resolved the "Conflicting files"

FYI @Vanethos

**Comment by KevinTheGray** [comment]
@brunobowden Thanks for the feedback!  I'll make those changes.   I'm actually not sure why the README got formatted like that.. I switched to VSCode recently and I think it autoformatted the file like that for some reason.  I'll get that reverted.

**Comment by brunobowden** [comment]
@KevinTheGray - thanks for doing a thoughtful job on this. I hope that this works for @cyanglaz.

File: packages/connectivity/connectivity/pubspec.yaml
Comment (KevinTheGray) [reply packages/connectivity/connectivity/pubspec.yaml]: updated to 0.5.0

File: packages/connectivity/connectivity/pubspec.yaml
Comment (KevinTheGray) [reply packages/connectivity/connectivity/pubspec.yaml]: reverted to double quotes

File: packages/connectivity/connectivity/CHANGELOG.md
Comment (KevinTheGray) [reply packages/connectivity/connectivity/CHANGELOG.md]: reverted this change

File: packages/connectivity/connectivity/CHANGELOG.md
Comment (KevinTheGray) [reply packages/connectivity/connectivity/CHANGELOG.md]: fixed

File: packages/connectivity/connectivity/CHANGELOG.md
Comment (KevinTheGray) [reply packages/connectivity/connectivity/CHANGELOG.md]: Added link

File: packages/connectivity/connectivity/CHANGELOG.md
Comment (KevinTheGray) [reply packages/connectivity/connectivity/CHANGELOG.md]: updated 0.5.0

File: packages/connectivity/connectivity/ios/Classes/FLTConnectivityPlugin.m
Comment (brunobowden) [inline packages/connectivity/connectivity/ios/Classes/FLTConnectivityPlugin.m:109]: I would put the code back in the old position. By moving it, there's no easily ability to see if the code has been edited. Leaving it in the old position saves a single `#ndef` which is a very modest cost.

File: packages/connectivity/connectivity/ios/Classes/FLTConnectivityPlugin.m
Comment (brunobowden) [inline packages/connectivity/connectivity/ios/Classes/FLTConnectivityPlugin.m:77]: Makes sense to move this since it's a very local change

File: packages/connectivity/connectivity/example/lib/main.dart
Comment (brunobowden) [inline packages/connectivity/connectivity/example/lib/main.dart:160]: Is this catch needed? I thought that the exception could only happen for the `getWifiName` and `getWifiBSSID` methods.... while `getWifiIP` would continue working as normal?

File: packages/connectivity/connectivity/README.md
Comment (brunobowden) [inline packages/connectivity/connectivity/README.md:108]: Append
```
When disabled, the `getWifiName` and `getWifiBSSID` methods will throw a `MissingPluginException` error.
```

File: packages/connectivity/connectivity/README.md
Comment (brunobowden) [inline packages/connectivity/connectivity/README.md:119]: The location code doesn't matter, you really care about the location permission. This macro is also meant to match the "No Location Permission" in the README.
```
NO_CONNECTIVITY_LOCATION_PERMISSION
```

**Review by brunobowden — APPROVED** [review]
@KevinTheGray - this is looking in good shape. You have my approval based on addressing these comments. Obviously this is a first step to getting this merged.

**Comment by brunobowden** [comment]
Change title in commit and GitHub. Again, assume that Android could adopt the same policy, so this is the basis for a cross platform "Don't make me require Location Permissions" config.

Suggested Title: `[connectivity] No Location Permission preprocessor flag`

**Comment by KevinTheGray** [comment]
Thanks again! I'll go through this and update asap

File: packages/connectivity/connectivity/CHANGELOG.md
Comment (KevinTheGray) [reply packages/connectivity/connectivity/CHANGELOG.md]: updated

File: packages/connectivity/connectivity/README.md
Comment (KevinTheGray) [reply packages/connectivity/connectivity/README.md]: done

File: packages/connectivity/connectivity/README.md
Comment (KevinTheGray) [reply packages/connectivity/connectivity/README.md]: done

File: packages/connectivity/connectivity/README.md
Comment (KevinTheGray) [reply packages/connectivity/connectivity/README.md:120]: done

File: packages/connectivity/connectivity/README.md
Comment (KevinTheGray) [reply packages/connectivity/connectivity/README.md:127]: thanks!

File: packages/connectivity/connectivity/README.md
Comment (KevinTheGray) [reply packages/connectivity/connectivity/README.md]: done

File: packages/connectivity/connectivity/README.md
Comment (KevinTheGray) [reply packages/connectivity/connectivity/README.md]: updated to NO_LOCATION_PERMISSION_CONNECTIVITY

File: packages/connectivity/connectivity/example/lib/main.dart
Comment (KevinTheGray) [reply packages/connectivity/connectivity/example/lib/main.dart]: you're right! took that out

File: packages/connectivity/connectivity/ios/Classes/FLTConnectivityPlugin.m
Comment (KevinTheGray) [reply packages/connectivity/connectivity/ios/Classes/FLTConnectivityPlugin.m:118]: sounds good

File: packages/connectivity/connectivity/ios/Classes/FLTConnectivityPlugin.m
Comment (KevinTheGray) [reply packages/connectivity/connectivity/ios/Classes/FLTConnectivityPlugin.m]: moved all blocks as close as possible to their old positions

File: packages/connectivity/connectivity/CHANGELOG.md
Comment (brunobowden) [inline packages/connectivity/connectivity/CHANGELOG.md:3]: It took them 45 years but Microsoft Word has finally spoken about 2 spaces after a period: https://www.cnn.com/2020/04/27/tech/microsoft-word-two-spaces-trnd/index.html

As a general principle, you should maintain the same convention that's there in the existing code.

File: packages/connectivity/connectivity/README.md
Comment (brunobowden) [inline packages/connectivity/connectivity/README.md:108]: Two space, repeated twice: https://www.cnn.com/2020/04/27/tech/microsoft-word-two-spaces-trnd/index.html

Comment (brunobowden) [inline packages/connectivity/connectivity/README.md:108]: Setting the usage description strings isn't what the developer cares about. What they care about is not using location permissions:

```suggestion
If you don't wish to use location permissions on iOS, you also have the option to disable the code that uses these permissions. This may be useful if you don't wish to specify permissions that you have no intention of using and only want to use the plugin for checking the connection status of the device. When disabled, the `getWifiName` and `getWifiBSSID` methods will throw a `MissingPluginException` error.
```

File: packages/connectivity/connectivity/example/lib/main.dart
Comment (brunobowden) [inline packages/connectivity/connectivity/example/lib/main.dart:127]: This should be part of a debug message and not the wifiName, which someone could interpret as a valid name. But then they do use it in this way on L146.

So I vote to keep it as is and see what they say in review.

**Review by brunobowden — APPROVED** [review]
Slights nits but approval from me.

File: packages/connectivity/connectivity/CHANGELOG.md
Comment (KevinTheGray) [reply packages/connectivity/connectivity/CHANGELOG.md]: Good to know!  updated

File: packages/connectivity/connectivity/README.md
Comment (KevinTheGray) [reply packages/connectivity/connectivity/README.md]: fixed

Comment (KevinTheGray) [reply packages/connectivity/connectivity/README.md]: Good point, updated

**Comment by brunobowden** [comment]
@KevinTheGray - see the failure here. Nice to get all the checks passing ready for review:
https://github.com/flutter/plugins/pull/2771/checks?check_run_id=1207080912

**Comment by brunobowden** [comment]
@KevinTheGray - still two failures: both CRC checks. My first suggestion would be to rebase and hope that someone else has fixed this. I expect this is causing other PRs to fail too:

https://github.com/flutter/plugins/pull/2771/checks?check_run_id=1248587802
https://github.com/flutter/plugins/pull/2771/checks?check_run_id=1248587782

```
Downloading ios-release tools...                                    8.2s
/Users/anka/flutter/bin/cache/artifacts/engine/ios-release/Flutter.framework.zip  bad CRC f7b6ef7b  (should be d52a7539)
Downloading ios-release tools...                                    7.9s
/Users/anka/flutter/bin/cache/artifacts/engine/ios-release/Flutter.framework.zip  bad CRC 56124e25  (should be d52a7539)
```

**Comment by KevinTheGray** [comment]
@brunobowden Just pulled in master and it's currently passing now.  I'll keep an eye on it, seems like the CI will periodically re-trigger and sometimes fail even if I haven't pushed any new changes

**Comment by brunobowden** [comment]
Thanks @KevinTheGray. I'd just like to keep it warm for when @cyanglaz gets a chance to look at it.

**Comment by brunobowden** [comment]
@cyanglaz @bparrishMines - perfectly respect the decision on making a separate plugin. It does make the most long term sense. Any ideas on the time frame?

@KevinTheGray - good working with you on the PR. Though this may not make the light of day, better things will come in its' place.

**Comment by KevinTheGray** [comment]
@brunobowden thanks for the reviews!  The separate plugin approach will definitely keep things simpler going forward. In the meantime i can just use my branch until the separation is complete.

**Comment by brunobowden** [comment]
@KevinTheGray - how do you use your branch? Is it published as a plugin? I might have an interest in using it for the WHO app depending on the timeline for wifi_info_flutter plugin.

**Comment by KevinTheGray** [comment]
You can add this to your pubspec.yaml dependencies
```
connectivity:
    git:
      url: https://github.com/KevinTheGray/plugins
      ref: kg-connectivity-ios-permissions
      path: packages/connectivity/connectivity
```
you can also lock it to a certain commit if desired: https://dart.dev/tools/pub/dependencies

**Comment by anotherglitchinthematrix** [comment]
@KevinTheGray it says "Package doesn't exist (the pubspec for connectivity 0.5.0 from git has version 0.4.10"

**Comment by KevinTheGray** [comment]
@anotherglitchinthematrix that's strange, can you share your `puspec.yaml`?  I just created a new flutter project to test it out, and it seems to be working for me.

**Comment by anotherglitchinthematrix** [comment]
@KevinTheGray Sure, thank you.
```
version: 0.5.0+82

environment:
  sdk: ">=2.6.0 <3.0.0"

dependencies:
  flutter:
    sdk: flutter
  provider: ^4.3.2
  hive: ^1.4.3
  http: ^0.12.2
  xml: ^4.2.0
  html_unescape: ^1.0.1+3
  firebase_core: ^0.5.0
  firebase_crashlytics: ^0.2.1+1
  firebase_analytics: ^5.0.16
  firebase_auth: ^0.18.1+1
  cloud_firestore: ^0.14.1+2
  firebase_messaging: ^7.0.2
  collection: ^1.14.13
  intl: ^0.16.1
  share: ^0.6.5+2
  url_launcher: ^5.7.2
  path_provider: ^1.6.18
  connectivity: ^0.4.9+3  # I revert it to original.
  dio: ^3.0.10
  dio_http_cache: ^0.2.11

dev_dependencies:
  flutter_test:
    sdk: flutter
  hive_flutter: ^0.3.1
  hive_generator: ^0.7.1
  build_runner: ^1.10.1
  flutter_launcher_icons: ^0.8.0

flutter_icons:
  android: true
  ios: true
  image_path: "asset/icon/icon.png"
  adaptive_icon_background: "asset/icon/icon.png"
  adaptive_icon_foreground: "asset/icon/icon.png"
```

**Pull request closed without merging**
